core/src/smartcontracts/isi/asset.rs

```diff
@@ -243,45 +272,57 @@ pub mod isi {
                 wsv,
                 <Self as AssetInstructionInfo>::EXPECTED_VALUE_TYPE,
             )?;
-            wsv.modify_asset(&asset_id, |asset| {
+            let quantity = {
+                let account = wsv.account_mut(&asset_id.account_id)?;
+                let asset = account
+                    .assets
+                    .get_mut(&asset_id)
+                    .ok_or_else(|| FindError::Asset(asset_id.clone()))?;
                 let quantity: &mut Self = asset
                     .try_as_mut()
                     .map_err(eyre::Error::from)
                     .map_err(|e| Error::Conversion(e.to_string()))?;
                 *quantity = quantity
                     .checked_sub(burn.object)
                     .ok_or(MathError::NotEnoughQuantity)?;
-                #[allow(clippy::float_arithmetic)]
-                wsv.metric_tx_amounts
-                    .set(wsv.metric_tx_amounts.get() + (*quantity).into_metric());
-                wsv.metric_tx_amounts_counter
-                    .set(wsv.metric_tx_amounts_counter.get() + 1);
+                let quantity = *quantity;
+                if asset.value.is_zero_value() {
+                    assert!(account.remove_asset(&asset_id).is_some());
+                }
+                quantity
+            };
+
+            // TODO: shouldn't we add only increment?
```

Comment:
Could you explain the issue more?

Reply:
I've linked issue and rephrased. Is it enough?

Reply:
Yup, thanks.